crates/accelerate/src/gate_direction.rs

```diff
@@ -142,8 +159,471 @@ where
+            DAGNode { node: None },
+        ),
+    )?;
+
+    dag.has_calibration_for(py, dag_op_node.borrow(py))
+}
+
+// Return a replacement DAG for the given standard gate in the supported list
+// TODO: optimize it by caching the DAGs of the non-parametric gates and caching and
+// mutating upon request the DAGs of the parametric gates
+fn replace_dag(
+    py: Python,
+    std_gate: StandardGate,
+    inst: &PackedInstruction,
+) -> PyResult<DAGCircuit> {
+    let replacement_dag = match std_gate {
+        StandardGate::CXGate => cx_replacement_dag(py),
+        StandardGate::ECRGate => ecr_replacement_dag(py),
+        StandardGate::CZGate => cz_replacement_dag(py),
+        StandardGate::SwapGate => swap_replacement_dag(py),
+        StandardGate::RXXGate => rxx_replacement_dag(py, inst.params_view()),
+        StandardGate::RYYGate => ryy_replacement_dag(py, inst.params_view()),
+        StandardGate::RZZGate => rzz_replacement_dag(py, inst.params_view()),
+        StandardGate::RZXGate => rzx_replacement_dag(py, inst.params_view()),
+        _ => panic!("Mismatch in supported gates assumption"),
+    };
+
+    replacement_dag
+}
+
+//###################################################
+// Utility functions to build the replacement dags
+//
+// TODO: replace this once we have a Rust version of QuantumRegister
+#[inline]
+fn add_qreg(py: Python, dag: &mut DAGCircuit, num_qubits: u32) -> PyResult<Vec<Qubit>> {
+    let qreg = imports::QUANTUM_REGISTER
+        .get_bound(py)
+        .call1((num_qubits,))?;
+    dag.add_qreg(py, &qreg)?;
+    let mut qargs = Vec::new();
+
+    for i in 0..num_qubits {
+        let qubit = qreg.call_method1(intern!(py, "__getitem__"), (i,))?;
+        qargs.push(dag.qubits().find(&qubit).expect("Qubit should stored"));
+    }
+
+    Ok(qargs)
+}
+
+#[inline]
+fn apply_operation_back(
+    py: Python,
+    dag: &mut DAGCircuit,
+    gate: StandardGate,
+    qargs: &[Qubit],
+    param: Option<SmallVec<[Param; 3]>>,
+) -> PyResult<()> {
+    dag.apply_operation_back(
+        py,
+        PackedOperation::from_standard(gate),
+        qargs,
+        &[],
+        param,
+        ExtraInstructionAttributes::new(None, None, None, None),
+        #[cfg(feature = "cache_pygates")]
+        None,
+    )?;
+
+    Ok(())
+}
+
+fn cx_replacement_dag(py: Python) -> PyResult<DAGCircuit> {
+    let new_dag = &mut DAGCircuit::new(py)?;
+    let qargs = add_qreg(py, new_dag, 2)?;
+    let qargs = qargs.as_slice();
+
+    apply_operation_back(py, new_dag, StandardGate::HGate, &[qargs[0]], None)?;
+    apply_operation_back(py, new_dag, StandardGate::HGate, &[qargs[1]], None)?;
+    apply_operation_back(
+        py,
+        new_dag,
+        StandardGate::CXGate,
+        &[qargs[1], qargs[0]],
+        None,
+    )?;
+    apply_operation_back(py, new_dag, StandardGate::HGate, &[qargs[0]], None)?;
+    apply_operation_back(py, new_dag, StandardGate::HGate, &[qargs[1]], None)?;
+
+    Ok(new_dag.clone())
+}
```

Comment:
Could we maybe optimize these circuit building functions by having them work more like `OnceCell`? Where you could create a static instance that is initialized just once, you can then just call clone each time you need a replacement `DAGCircuit`. For this you could use `GILOnceCell`, and have these function calls just call on `.get_or_init(py, |_| )` 

Here's an example of how it could work:
```rust

static FOO_DAG: GILOnceCell<PyResult<DAGCircuit>> = GILOnceCell::new();

fn foo_replacement_dag(py: Python) -> PyResult<DAGCircuit> {
    FOO_DAG.get_or_init(py, || -> PyResult<DAGCircuit> {
        let mut dag: DAGCircuit = DAGCircuit::new(py);
        // Not sure if you can call `add_qreg` from within the closure.
        let qargs = add_qreg(py, new_dag, 2)?;
        let qargs = qargs.as_slice();
        // not sure if you can call `apply_operation_back` from inside the closure.
        apply_operation_back(py, &mut dag, StandardGate::XGate, &[qargs[0]], None)?;
        Ok(dag)
    }).cloned()
}
```
This ensures that the circuit is built one and that whenever we call on this function, we just clone the circuit.

This change is merely a suggestion since #13335 makes improvements on chained additions to the `DAGCircuit` and these changes could be made in a follow up 🙃.

Reply:
As said in https://github.com/Qiskit/qiskit/pull/13069#discussion_r1820706305 this is definitely an optimization opportunity I would like to explore.  I am actually exploring already some caching implementation, similar to what you suggested, namely generating a replacement DAG (on-demand) once and cloning when needed. This applies to non-parameterized gates only since replacements for parameterized gates (e.g. RZX) depend on the actual parameter. It also includes caching a template DAG for RZX and then when an RZX replacement is needed replacing the RZX node with a concrete parametrically-bound RZX node.  Then I also want to look at avoiding calling `.clone()` , which I believe could give even more improvement, but this will require more refactoring. But bottom-line I'm afraid I will not have time to properly explore and benchmark more optimization for the 1.3 release, so I think we should have this in a new PR. And obviously there is a risk of jut trying to micro optimize without too much benefit, but again, I'll need more time to benchmark in detail.